File: 04_Binary_Search/05_Search_in_Rotated_Sorted_Array.java
class Solution {
    public int search(int[] nums, int target) {
        int l = 0, r = nums.length - 1;
        int mid = -1;

        while (l <= r) {
            mid = l + (r - l) / 2;
            if (nums[mid] == target)
                return mid;

            // MID IS IN 'LEFT' SORTED PORTION
            if (nums[l] <= nums[mid]) {
                if (target > nums[mid] || target < nums[l])
                    l = mid + 1;
                // target < nums[mid] || target > nums[l]
                else
                    r = mid - 1;
            }
            // MID IS IN 'RIGHT' SORTED PORTION
            else {
                if (target < nums[mid] || target > nums[r])
                    r = mid - 1;
                // target > nums[mid] or target < nums[r]
                else
                    l = mid + 1;
            }
        }

        return -1;
    }
}

File: 04_Binary_Search/05_Search_in_Rotated_Sorted_Array.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: int
        """
        l, r = 0, len(nums) - 1

        while l <= r:

            mid = l + (r - l) // 2

            if nums[mid] == target:
                return mid
            
            # MID IS IN THE 'LEFT' SORTED PORTION OF THE ARRAY
            if nums[mid] >= nums[l]:
                if target > nums[mid] or target < nums[l]:
                    l = mid + 1
                else:
                    r = mid - 1
            # MID IS IN THE 'RIGHT' SORTED PORTION OF THE ARRAY
            else:
                if target < nums[mid] or target > nums[r]:
                    r = mid - 1
                else:
                    l = mid + 1
            
        return -1